cunumeric/module.py

```diff
@@ -1190,6 +1203,9 @@ def transpose(a: ndarray, axes: Optional[list[int]] = None) -> ndarray:
     --------
     Multiple GPUs, Multiple CPUs
     """
+    if a is None:
+        return ndarray((), dtype=object)  # type: ignore [unreachable]
+
```

Comment:
TL;DR: NumPy does not document that it accepts `None` in this function. My suggested approach for handling undocumented inputs in general is that we can do whatever we want. In this particular case, I think we should not bother.

We are already doing a good job in this case mirroring NumPy's *documented* behavior in our function signature[^1]. NumPy also accepts `None`, but that is not noted on the NumPy function signature or docstring. Adding explicit code to handle this undocumented/unexpected case ends up looking weird (and mypy agrees).

If we do decide to accept an undocumented value like `None`, then let's be systematic and actually include it in our signature, instead of adding "unreachable" annotations.

In this particular case, I think we actually don't want to accept `None`, because it appears NumPy handles it by internally converting it to an `object`-type array somewhere, which is not a dtype that's currently supported by cuNumeric, so doing anything with this result going forward is likely to fail miserably.

For context, my overall suggested approach is as follows:

| NumPy accepts a value? | what should cuNumeric do? |
| ------------- | ------------- |
| yes, documented | accept, or throw `NotImplementedError` |
| yes, undocumented | whatever goes |
| no, and checks explicitly[^2] | reject, raise same exception |
| no, and doesn't check[^3] | whatever goes |

[^1]: We should come up with a better way to "derive" the actual function signature, after `@add_boilerplate` is applied, but that's a question for another time. In this case `@add_boilerplate` transforms the "internal" signature of `(a: cn.ndarray, axes: None | list[int]) -> cn.ndarray` to what the application actually sees: `(a: npt.ArrayLike | cn.ndarray | SupportsLegateDataInterface, axes: None | list[int]) -> ndarray`.

[^2]: We can usually tell because it throws an "intentional" error, like `ValueError`, `TypeError`, `IndexError`, `LinAlgError`.

[^3]: We can usually tell because it fails at an "arbitrary" point in the code, when the code tries to use an incompatible value in a way that it implicitly assumes it can accept, e.g. throwing `AttributeError` when it tries to get an attribute from a `None` value.

Reply:
>  My suggested approach for handling undocumented inputs in general is that we can do whatever we want. 

OK but in that case we should probably remove relevant `xfail` tests altogether then, since is it neither expected to fail nor to pass

> If we do decide to accept an undocumented value like None, then let's be systematic and actually include it in our signature, instead of adding "unreachable" annotations.

I would be :-1: on adding type values that are only ever expected to fail, so I'd agree with the above and just not handle them at all. I guess the problem arises when Numpy itself is vague or not completely clear. 

> In this particular case, I think we actually don't want to accept None, because it appears NumPy handles it by internally converting it to an object-type array somewhere, which is not a dtype that's currently supported by cuNumeric, so doing anything with this result going forward is likely to fail miserably.

That seems to be a case for `NotImplemented` then?

Reply:
> I guess the problem arises when Numpy itself is vague or not completely clear.

For instance, the docstring for `np.transpose` just says
```
a : array_like
    Input array.
```
`ArrayLike` includes ["the likes of Scalars"](https://numpy.org/devdocs/reference/typing.html#numpy.typing.ArrayLike). Does that  mean None too? Hard to tell tbh

Reply:
> If we do decide to accept an undocumented value like None, then let's be systematic and actually include it in our signature, instead of adding "unreachable" annotations. 

Also regarding this, Numpy docs also state:

> Trying to describe the full range of possibilities statically would result in types that are not very helpful. For that reason, the typed NumPy API is often stricter than the runtime NumPy API.

Which is more or less what I was trying to do in this PR. I guess mostly I just believe that if `cunumeric` is intended to be a drop-in replacement for numpy, then that means truly cloning as much as possible, "undefined" behavior and all.